### PR TITLE
feat: add per strategy apr, apy and debt ratio

### PR DESCRIPTION
**Pull request by x48115**

### CODE REVIEW

**Review by dudesahn — COMMENTED** [review]
Would it make more sense to put the strategy-apy calcs somewhere in the APY directory? like, maybe in the apy/v2.py file?

File: yearn/v2/strategies.py [hunk 1]
```diff
@@ -88,6 +91,7 @@ def process_events(self, events):
                 block = event.block_number
                 logger.debug("%s harvested on %d", self.name, block)
                 self._harvests.append(block)
+                self._harvests_data.append({"block_number": block, "event": event})
```
Comment (banteg) [inline yearn/v2/strategies.py:94]: isn't it redundant? you can get the block number from `event.block_number`.

File: yearn/v2/strategies.py [hunk 2]
```diff
@@ -96,8 +100,96 @@ def load_harvests(self):
...
+    @property
+    def harvests_data(self):
+        if len(self._harvests) == 0:
+            self.load_harvests()
+        sorted_harvests = sorted(self._harvests_data, key=lambda k: k['block_number'], reverse=True)
+        return sorted_harvests
+
+    @property
+    def debt_ratio(self):
+        return self.vault.vault.strategies(self.strategy.address).dict()['debtRatio'] / 1e4
```
Comment (banteg) [inline yearn/v2/strategies.py:116]: you can access struct items without converting to dict, contracts also work where address is expected
```suggestion
        return self.vault.vault.strategies(self.strategy)['debtRatio'] / 1e4
```

**Review by banteg — APPROVED** [review]

File: yearn/v2/strategies.py [hunk 3]
```diff
@@ -96,8 +100,96 @@ def load_harvests(self):
...
+        if profitable_harvests_count < 2:
+            return StrategyApy(0, 0)
+
+        # Not enough data
+        have_enough_data = second_latest_harvest_with_profit_idx + 1 < len(harvests)
+        if have_enough_data == False:
+            return StrategyApy(0, 0)
+
+        # Latest profitable harvest
+        latest_harvest_with_profit_current = harvests[latest_harvest_with_profit_idx]
```
Comment (mariuspod) [inline yearn/v2/strategies.py:144]: this part is pretty hard to read and debug imo. Consider adding some helper methods and return a tuple or a dataclass.
E.g.
```
first_harvest_current, second_harvest_current, duration_current = _get_profitable_harvests_with_duration(harvests, idx+1)
first_harvest_previous, second_harvest_previous, duration_previous = _get_profitable_harvests_with_duration(harvests, idx)
```

Another good option would be to use a dataclass for the return value and then use it for the apr calculations so it becomes:
```
current = _get_harvests_with_duration(idx+1)
previous = _get_harvests_with_duration(idx)
```

then further below it becomes:
```
current_apr = current.profit / ...
```

File: yearn/v2/strategies.py [hunk 4]
```diff
@@ -96,8 +100,96 @@ def load_harvests(self):
...
+        strategy_performance_fee = vault.strategies(strategy.address).dict()['performanceFee'] / 1e4
+        vault_performance_fee = vault.performanceFee() / 1e4
+        performance_fee = vault_performance_fee + strategy_performance_fee
+        management_fee = vault.managementFee() / 1e4 if hasattr(vault, "managementFee") else 0
+        
+        # Subtract fees
+        apr_minus_fees = apr * (1 - performance_fee) - management_fee
+        
+        # Convert APR to annualized estimated APY
+        net_apy = (1 + (apr_minus_fees / 365)) ** 365 - 1
```
Comment (mariuspod) [inline yearn/v2/strategies.py:189]: Not trying to be pedantic here, I'm just wondering why we're using here 365 if the `SECONDS_IN_YEAR` constant uses 365.25 as a base. Maybe it's not an issue and a common economic agreement to use 365 for apy calculations but it looks slightly off and both values should either use 365 or 365.25 imo.

**Review by mariuspod — APPROVED** [review]
the requested changes for improving code readability are optional but appreciated :sweat_smile:

File: yearn/v2/strategies.py (hunk 4 above)
Comment (dudesahn) [reply yearn/v2/strategies.py:189]: Alternatively, we could instead use daily for fantom, and weekly for Ethereum. But I'm not quite sure how to do this in exporter; is there a way to read what network a contract is on?

Comment (mariuspod) [reply yearn/v2/strategies.py:189]: @dudesahn the exporter is running within a given brownie network that you provide on startup. So basically all contracts seen within one exporter process are scoped to this one network. As you already said the current network can be checked with `chain.id` and we have some constants in `yearn/networks.py` which could be used like so:

https://github.com/yearn/yearn-exporter/blob/master/yearn/prices/magic.py#L42

**Pull request closed without merging**